addons/account/tests/test_tax_report.py

```diff
@@ -178,3 +178,72 @@ def test_tax_report_change_country(self):
                 # Tags already exist since 'copied_report_1' belongs to 'test_country_2'
                 for tag in line_tags:
                     self.assertIn(tag, country_2_tags_after_change, "The tax_tags expressions sharing their tags with other report should not receive new tags since they already exist.")
+
+    def test_unlink_report_line_tags_used_by_amls(self):
+        """
+        Deletion of a report line whose tags are still referenced by an aml should archive tags and not delete them.
+        """
+        tag_name = "55b"
+        tax_report_line = self._create_basic_tax_report_line(self.tax_report_1, "Line 55 bis", tag_name)
+        test_tag = tax_report_line.expression_ids._get_matching_tags().filtered(lambda tag: not tag.tax_negate)
+        test_tax = self.env['account.tax'].create({
+            'name': "Test tax",
+            'amount_type': 'percent',
+            'amount': 25,
+            'country_id': self.tax_report_1.country_id.id,
+            'type_tax_use': 'sale',
+            'invoice_repartition_line_ids': [
+                (0, 0, {'factor_percent': 100, 'repartition_type': 'base'}),
+                (0, 0, {
+                    'factor_percent': 100,
+                    'repartition_type': 'tax',
+                    'tag_ids': [Command.link(test_tag.id)],
+                }),
+            ],
+            'refund_repartition_line_ids': [
+                (0, 0, {'factor_percent': 100, 'repartition_type': 'base'}),
+                (0, 0, {'factor_percent': 100, 'repartition_type': 'tax'}),
+            ],
+        })
+
+        # Make sure the fiscal country allows using this tax directly
+        self.env.company.account_fiscal_country_id = self.test_country_1
+
+        test_invoice = self.env['account.move'].create({
+            'move_type': 'out_invoice',
+            'partner_id': self.partner_a.id,
+            'date': '1992-12-22',
+            'invoice_line_ids': [
+                (0, 0, {'quantity': 1, 'price_unit': 42, 'tax_ids': [Command.set([test_tax.id])]}),
+            ],
+        })
+        test_invoice.action_post()
+
+        tax_report_line.unlink()
+        plus_tags_after = self._get_tax_tags(self.test_country_1, tag_name=tag_name, active_test=False).filtered(lambda tag: tag.name[:1] == '+')
+        self.assertEqual(len(plus_tags_after), 1, "Unlinking a report line whose tags are used on move lines should keep the right sign.")
```

Comment:
Error does not really mean anything; make it clearer: 
Unlinking a tax_tags expression should keep the tag if it was used on move lines, and unlink it otherwise. 

You also need to check that - sign got removed. So to me the assert could directly check that tax_report_line.expression_ids._get_matching_tags().mapped(tax_negate) == [False]

Reply:
You can't use tax_report_line directly here since it's unlinked but I get the idea :+1: